src/bin/commands/run.rs

```diff
@@ -0,0 +1,108 @@
+use clap::Args as ClapArgs;
+use cross::config::Config;
+use cross::shell::{MessageInfo, Verbosity};
+use cross::SafeCommand;
+use cross::{
+    cargo_metadata_with_args, cli::Args, docker, rustc, setup, toml, CargoVariant, CrossSetup,
+    Target,
+};
+use eyre::Context;
+
+#[derive(ClapArgs, Debug)]
+pub struct Run {
+    /// Provide verbose diagnostic output.
+    #[clap(short, long)]
+    pub verbose: bool,
+    /// Do not print cross log messages.
+    #[clap(short, long)]
+    pub quiet: bool,
+    /// Coloring: auto, always, never
+    #[clap(long)]
+    pub color: Option<String>,
+    /// Container engine (such as docker or podman).
+    #[clap(long)]
+    pub engine: Option<String>,
+
+    #[clap(short, long)]
+    pub target: String,
+
+    #[clap(short, long)]
+    pub command: String,
+}
+
+impl Run {
+    pub fn run(&self, engine: docker::Engine, msg_info: &mut MessageInfo) -> cross::Result<()> {
+        let target_list = rustc::target_list(&mut Verbosity::Quiet.into())?;
+        let target = Target::from(&self.target, &target_list);
+
+        let cwd = std::env::current_dir()?;
+        let host_version_meta = rustc::version_meta()?;
+
+        let args = Args {
+            cargo_args: vec![],
+            rest_args: vec![],
+            subcommand: None,
+            channel: None,
+            target: Some(target.clone()),
+            features: vec![],
+            target_dir: None,
+            manifest_path: None,
+            version: false,
+            verbose: if self.verbose { 1 } else { 0 },
+            quiet: self.quiet,
+            color: self.color.clone(),
+        };
+
+        if let Some(metadata) = cargo_metadata_with_args(None, Some(&args), msg_info)? {
+            let CrossSetup { toolchain, .. } =
+                match setup(&host_version_meta, &metadata, &args, target_list, msg_info)? {
+                    Some(setup) => setup,
+                    _ => {
+                        eyre::bail!("Error: cannot setup cross environment");
+                    }
+                };
+
+            let toml = toml(&metadata, msg_info)?;
+            let config = Config::new(toml);
+
+            let image = match docker::get_image(&config, &target, false) {
+                Ok(i) => i,
+                Err(err) => {
+                    msg_info.warn(&err)?;
+                    eyre::bail!("Error: {}", &err);
+                }
+            };
+
+            let image = image.to_definite_with(&engine, msg_info);
+
+            let paths = docker::DockerPaths::create(&engine, metadata, cwd, toolchain, msg_info)?;
+            let options =
+                docker::DockerOptions::new(engine, target, config, image, CargoVariant::None, None);
+
+            let command = SafeCommand::new("sh");
+            let mut args = vec![String::from("-c")];
```

Comment:
I agree that sh may not be available, but we already use sh all around the codebase and inside of docker like in [1](https://github.com/cross-rs/cross/blob/main/src/bin/commands/containers.rs#L441) [2](https://github.com/cross-rs/cross/blob/main/src/docker/remote.rs#L798) [3](https://github.com/cross-rs/cross/blob/main/src/docker/remote.rs#L56) [4](https://github.com/cross-rs/cross/blob/main/src/docker/remote.rs#L188) [5](https://github.com/cross-rs/cross/blob/main/src/docker/shared.rs#L892)

And all cargo/xargo/zig commands are already proxied through sh, can see it [here](https://github.com/cross-rs/cross/blob/main/src/docker/shared.rs#L1085) or if you run cross in verbose mode will print the docker command that is used and you can see that we are using sh to inject the sysroot bin in the PATH.

Also I think is handy to run the command in a shell so you can use pipes and other shell elements, but is also true that you could always do `cross-util run --target <target> -- sh -c "mycommand | withpipes"` but the user will need to scape then what's inside `"`. So even if we run the command "directly" it will be run through sh.

I will maintain the sh for now until we can configure the shell to use in Cross.toml for each target/container. WDYT?

Reply:
Done, it was easier than expected and also really useful.